internal/app/proxyd/main.go

```diff
@@ -48,46 +52,70 @@ func main() {
 	// Start up with initial bootstrap config
 	go r.Bootstrap(bootstrapCtx)
 
-	// nolint: errcheck
-	go func() {
-		kubeconfig := "/etc/kubernetes/admin.conf"
-		if err = conditions.WaitForFilesToExist(kubeconfig).Wait(context.Background()); err != nil {
-			log.Fatalf("failed to find %s: %v", kubeconfig, err)
-		}
+	go waitForKube(r)
 
-		config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
-		if err != nil {
-			log.Fatalf("failed to read config %s: %v", kubeconfig, err)
-		}
+	var wg sync.WaitGroup
 
-		// Discover local non loopback ips
-		ips, err := pkgnet.IPAddrs()
-		if err != nil {
-			log.Fatalf("failed to get local address: %v", err)
+	wg.Add(1)
+	// Start up reverse proxy
+	go func(r *frontend.ReverseProxy, wg *sync.WaitGroup) {
+		defer wg.Done()
+		if err := r.Listen(":443"); err != nil {
+			log.Fatal(err)
 		}
-		if len(ips) == 0 {
-			log.Fatalf("no IP address found for local api server")
-		}
-		ip := ips[0]
-
-		// Overwrite defined host so we can target local apiserver
-		// and bypass the admin.conf host which is configured for proxyd
-		config.Host = ip.String() + ":6443"
-
-		clientset, err := kubernetes.NewForConfig(config)
+	}(r, &wg)
+
+	wg.Add(1)
+	// Start up gRPC server
```

Comment:
This wait group isn’t needed. ListenAndServe will block

Reply:
it's so we can launch both as goroutines and have a consistency across both listeners instead of doing
```
go r.Listen()
factory.ListenAndServe()
```

Reply:
What you posted ^ looks a lot simpler and easier to follow IMO. Not a blocker for me but what you posted above is the more common approach.